spec/abci++/v4.md

```diff
@@ -0,0 +1,200 @@
+        step_p ← precommit
+    }
+}
+```
+
+### Receiving enough prevotes to precommit
+
+The following code is ran upon receiving 2f + 1 prevotes for the same block
+
+```go
+upon ⟨PROPOSAL, h_p, round_p, v, *⟩
+  from proposer(h_p, round_p)
+  AND 2f + 1 ⟨PREVOTE, h_p, vr, id(v)⟩ 
+while valid(v) ∧ step_p >= prevote for the first time do {
+    if (step_p = prevote) {
+        lockedValue_p ← v
+        lockedRound_p ← round_p
+        processProposalOutput ← Join ABCI.ProcessProposal(h_p, v)
+        // If the proposal is valid precommit as before.
+        // If it was invalid, precommit nil.
+        // Note that ABCI.ProcessProposal(h_p, v).accept is deterministic for all honest nodes.
+        precommit_value ← nil
+        if processProposalOutput.accept {
+            precommit_value ← id(v)
+        }
+        precommit_extension ← ABCI.ExtendVote(h_p, round_p, id(v))
+        broadcast ⟨PRECOMMIT, h_p, round_p, id(v), precommit_extension⟩
+        for evidence in processProposalOutput.evidence_list {
+            broadcast ⟨EVIDENCE, evidence⟩ 
+        }
+
+        step_p ← precommit
+    }
+    validValue_p ← v
+    validRound_p ← round_p
+}
+```
+
+And upon receiving 2f + 1 prevotes for nil:
+
+```go
+upon 2f + 1 ⟨PREVOTE, h_p, round_p, nil⟩ 
+  while step_p = prevote do {
+    // Join ABCI.ProcessProposal, and broadcast any evidence if it exists.
+    processProposalOutput ← Join ABCI.ProcessProposal(h_p, v)
+    for evidence in processProposalOutput.evidence_list {
+        broadcast ⟨EVIDENCE, evidence⟩ 
+    }
+
+    precommit_extension ← ABCI.ExtendVote(h_p, round_p, nil)
+    broadcast ⟨PRECOMMIT, h_p, round_p, nil, precommit_extension⟩
+    step_p ← precommit
+}
+```
+
+### Upon receiving a precommit
+
+Upon receiving a precommit `precommit`, we ensure that `ABCI.VerifyVoteExtension(precommit.precommit_extension) = true`
+before accepting the precommit. This is akin to how we check the signature on precommits normally, hence its not wrapped
+in the syntax of methods from the paper.
+
+### Precommit timeout
+
+Upon receiving 2f + 1 precommits, setup a timeout.
+
+```go
+upon 2f + 1 ⟨PRECOMMIT, h_p, vr, *⟩ for the first time, do {
+    schedule OnTimeoutPrecommit(h_p, round_p) to be executed after timeoutPrecommit(round_p)
+}
+```
+
+with OnTimeoutPrecommit defined as:
+
+```go
+def OnTimeoutPrecommit(height, round) {
+    if (height = h_p && round = round_p) {
+        StartRound(round_p + 1)
+    }
+}
+```
+
+### Upon Receiving 2f + 1 precommits
+
+The following code is ran upon receiving 2f + 1 precommits for the same block
+
+```go
+upon ⟨PROPOSAL, h_p, r, v, *⟩
+  from proposer(h_p, r)
+  AND 2f + 1 ⟨ PRECOMMIT, h_p, r, id(v)⟩ 
+  while decision_p[h_p] = nil do {
+    if (valid(v)) {
+        decision_p[h_p] ← v
+        h_p ← h_p + 1
+        reset lockedRound_p, lockedValue_p,validRound_p and validValue_p to initial values
+        ABCI.FinalizeBlock()
```

Comment:
What happens with precommit_extension upon commit?

Reply:
They are handled by PrepareProposal, the data may be included within the block (either in the header or tx data), or not included at all and just have been there as an intermediate step.